### ISSUE TITLE
Work with the latest PyTorch core

### ISSUE THREAD

**Issue by asuhan** [pytorch/xla#46]
We're currently pinned to https://github.com/pytorch/pytorch/commit/1ca0ec7299b9352123a07525320a04e39d657781. An attempt to use latest PyTorch breaks a significant number of tests: https://github.com/pytorch/xla/tree/asuhan/upgrade_pytorch.

Several issues need to be addressed to track latest PyTorch:

[ ] https://github.com/pytorch/pytorch/issues/14825
[ ] https://github.com/pytorch/pytorch/issues/14826
[ ] Figure out a way to address the constant `aten::size` results leaking out of graphs: https://gist.github.com/asuhan/9d4b21a5b0174caa0854bcd5f917afed. The probable cause for it is https://github.com/pytorch/pytorch/pull/14485. While this makes sense for generalized traces, it is hard to handle in a principled way in our XLA bridge, especially when the forward and backward traces aren't fused.

Last but not least, we need to live at head and know when breaking changes (for us) to PyTorch happen. Setting up CI will at least let us know.

**Comment by asuhan** [comment]
We'll have to work around https://github.com/pytorch/pytorch/issues/14826, if nothing else restore the incorrect old behavior since not updating the counter doesn't hurt training.

**Comment by asuhan** [comment]
https://github.com/pytorch/pytorch/issues/14825 is the expected behavior.

**Comment by apaszke** [comment]
@asuhan That looks reasonable. I do not understand the fused vs unfused difference, so can’t comment on that, but you can assume that shapes of gradients will match the shapes of intermediates in the forward pass, so there’s no need to guard on that

**Comment by asuhan** [comment]
@apaszke Fusion means a single graph which takes input and labels and returns gradients for all parameters. Having a single graph for forward and backward allows better compiler decisions (rematerialization for example) and lowers memory usage, which in turn allows running at higher batch sizes. Depending on hardware, this can be the difference between being able to fit the model or not.

**Comment by apaszke** [comment]
Makes sense, thanks for the explanation!

**Comment by asuhan** [comment]
We still need to setup CI and upstream the remaining `optimizer.py` changes at https://github.com/pytorch/xla/blob/master/pytorch.patch in some form, the rest is done.
